src/libmongoc/tests/unified/entity-map.c

```diff
@@ -0,0 +1,603 @@
+      goto done;
+   }
+
+   /* Build the client's URI. */
+   uri = test_framework_get_uri ();
+   if (use_multiple_mongoses && test_framework_is_mongos ()) {
+      /* TODO: Once CDRIVER-3285 is resolved, update this to no longer rely on
+       * the private URI API. */
+      if (*use_multiple_mongoses) {
+         if (!mongoc_uri_upsert_host_and_port (uri, "localhost:27017", error)) {
+            goto done;
+         }
+         if (!mongoc_uri_upsert_host_and_port (uri, "localhost:27018", error)) {
+            goto done;
+         }
+      } else {
+         const mongoc_host_list_t *hosts;
+
+         hosts = mongoc_uri_get_hosts (uri);
+         if (hosts->next) {
+            test_set_error (error,
+                            "useMultiMongoses is false, so expected single "
+                            "host listed, but got: %s",
+                            mongoc_uri_get_string (uri));
+            goto done;
+         }
+      }
+   }
+
+   if (uri_options) {
+      /* Apply URI options. */
+      if (!uri_apply_options (uri, uri_options, error)) {
+         goto done;
+      }
+   }
+
+   client = mongoc_client_new_from_uri (uri);
+   entity->value = client;
+   callbacks = mongoc_apm_callbacks_new ();
+
+   if (observe_events) {
+      BSON_FOREACH (observe_events, iter)
+      {
+         const char *event_type = bson_iter_utf8 (&iter, NULL);
+
+         if (0 == strcmp (event_type, "commandStartedEvent")) {
+            mongoc_apm_set_command_started_cb (callbacks, command_started);
+         } else if (0 == strcmp (event_type, "commandFailedEvent")) {
+            mongoc_apm_set_command_failed_cb (callbacks, command_failed);
+         } else if (0 == strcmp (event_type, "commandSucceededEvent")) {
+            mongoc_apm_set_command_succeeded_cb (callbacks, command_succeeded);
+         } else {
+            test_set_error (error, "Unexpected event type: %s", event_type);
+            goto done;
+         }
+      }
+   }
+   mongoc_client_set_apm_callbacks (client, callbacks, entity);
+
+   ret = true;
+done:
+   mongoc_uri_destroy (uri);
+   bson_parser_destroy (parser);
+   mongoc_apm_callbacks_destroy (callbacks);
+   bson_destroy (uri_options);
+   bson_free (use_multiple_mongoses);
+   bson_destroy (observe_events);
+   if (!ret) {
+      entity_destroy (entity);
+      return NULL;
+   }
+   return entity;
+}
+
+entity_t *
+entity_database_new (entity_map_t *entity_map,
+                     bson_t *bson,
+                     bson_error_t *error)
+{
+   bson_parser_t *parser;
+   entity_t *entity;
+   const entity_t *client_entity;
+   char *client_id = NULL;
+   mongoc_client_t *client;
+   char *database_name = NULL;
+   bool ret = false;
+
+   entity = entity_new ("database");
+   parser = bson_parser_new ();
+   bson_parser_utf8 (parser, "id", &entity->id);
+   bson_parser_utf8 (parser, "client", &client_id);
+   bson_parser_utf8 (parser, "databaseName", &database_name);
```

Comment:
the spec lists another optional field `databaseOptions`

Reply:
Ah good catch. Added support for `databaseOptions` and `collectionOptions`.